src/Sparrow.Server/ByteString.cs

```diff
@@ -562,93 +562,105 @@ public void LowMemoryOver()
-                _minSize = size;
+            var pool = SegmentsPool;
+            if (pool.TryDequeue(out var segment) == false)
+                goto ALLOCATE;
 
-            var stack = SegmentsPool.Value;
+            Debug.Assert(segment != null, "The segment cannot be null.");
 
-            while (true)
+            if (segment.Size >= size)
             {
-                var current = stack.Head;
-                if (current == null)
-                    break;
-
-                if (Interlocked.CompareExchange(ref stack.Head, current.Next, current) != current)
-                    continue;
-
-                var segment = current.Value;
-                if (segment == null)
-                    continue;
+                if (segment.InUse.Raise() == false)
+                    goto ALLOCATE;
 
-                if (segment.Size >= size)
-                {
-                    if (!segment.InUse.Raise())
-                        continue;
-
-                    return segment;
-                }
-
-                // not big enough, so we'll discard it and create a bigger instance
-                // it will go into the pool afterward and be available for future use
-                segment.Dispose();
+                return segment;
             }
-
+            
+            // not big enough, so if the pool is half full we'll discard it and create a bigger instance
+            // it will go into the pool afterward and be available for future use.
+            // The idea is that we will get rid of small segments if half full is that we want the
+            // pool to only be half full on average. 
+            if (pool.Count >= SegmentPoolSize / 2 || pool.TryEnqueue(segment) == false)
+                segment.Dispose();
+                        
+            ALLOCATE:
+            // have to allocate it directly
             return new UnmanagedGlobalSegment(size);
         }
 
+        /// <summary>
+        /// Returns an UnmanagedGlobalSegment to the pool. If the segment is no longer in use and the pool is not full, 
+        /// it is added back for future reuse. 
+        /// 
+        /// If the pool is full, we attempt a fallback strategy:
+        /// 1. Dequeue another segment from the pool to compare sizes.
+        /// 2. Keep the larger segment (to better accommodate future allocations) and dispose of the smaller one.
+        /// 3. If re-enqueueing the chosen segment still fails, we dispose it to prevent memory leaks.
+        ///
+        /// This ensures that we maintain a healthy set of segments in the pool while also preventing indefinite growth or 
+        /// memory leaks if the pool cannot accept more segments.
+        /// </summary>
         public unsafe void Free(UnmanagedGlobalSegment memory)
         {
             ThrowIfNull<InvalidOperationException>(memory.Segment, "Attempt to return a memory segment that has already been disposed");
 
-            if (_minSize > memory.Size)
```

Comment:
Because we dont care about it anymore. The key is in the Allocate method, we will try to get a certain size, if we fail we allocate and return to the pool the one we got.